usr/share/rear/build/default/980_verify_rootfs.sh

```diff
@@ -1,18 +1,63 @@
+# to avoid issues like https://github.com/rear/rear/issues/1494
 
-# In case the filesystem is mounted 'noexec' we skip a chroot bash test
-tmp_filesystem=$(filesystem_name $ROOTFS_DIR)
+# In case the filesystem that contains the ROOTFS_DIR is mounted 'noexec' we cannot do the 'chroot' tests.
+# The filesystem_name function in linux-functions.sh returns the mountpoint (not a filesystem name like 'ext4'):
+local rootfs_dir_fs_mountpoint=$( filesystem_name $ROOTFS_DIR )
+if grep -qE '^\S+ '$rootfs_dir_fs_mountpoint' \S+ \S*\bnoexec\b\S* ' /proc/mounts ; then
+    # Intentionally failing here with Error to make it mandatory for ReaR to really check the rescue image.
+    # It is much more important to guarantee a runnable rescue image than to support noexec environments,
+    # cf. https://github.com/rear/rear/pull/1514#discussion_r140752346
+    Error "Cannot test if the ReaR recovery system is usable because $rootfs_dir_fs_mountpoint is mounted 'noexec'"
+fi
+Log "Testing that $ROOTFS_DIR contains a usable system"
 
-if grep -qE '^\S+ '$tmp_filesystem' \S+ \S*\bnoexec\b\S* ' /proc/mounts; then
-	LogPrint "WARNING: Filesystem $tmp_filesystem is mounted 'noexec', aborting chroot bash test"
-	Log "The above error means that we cannot test our chrooted environment because
-we need the 'exec' option set to the filesystem. One way to achieve this is
-by doing: mount -o remount,exec $tmp_filesystem"
-	return
+# The bash test ensures that we have a working bash in the ReaR recovery system:
+if ! chroot $ROOTFS_DIR /bin/bash -c true ; then
+    KEEP_BUILD_DIR=1
+    BugError "ReaR recovery system in '$ROOTFS_DIR' is broken: 'bash -c true' failed"
 fi
 
-# The chroot bash test ensures that we have a working bash on our rescue image
-if ! chroot $ROOTFS_DIR bash -c true ; then
-	KEEP_BUILD_DIR=1
-	BugError "ROOTFS_DIR '$ROOTFS_DIR' is broken, chroot bash test failed."
+# The ldd test ensures that for dynamically linked executables the required libraries are there.
+# The ldd test runs after the bash test because /bin/ldd is a bash script.
+# First test is 'ldd /bin/bash' to ensure 'ldd' works:
+Log "Testing 'ldd /bin/bash' to ensure 'ldd' works for the subsequent 'ldd' tests"
+if ! chroot $ROOTFS_DIR /bin/ldd /bin/bash 1>&2 ; then
+    KEEP_BUILD_DIR=1
+    BugError "ReaR recovery system in '$ROOTFS_DIR' is broken: 'ldd /bin/bash' failed"
+fi
+# Now test each binary (except links) with ldd and look for 'not found' libraries.
+# In case of 'not found' libraries for dynamically linked executables ldd returns zero exit code.
+# When running ldd for a file that is 'not a dynamic executable' ldd returns non-zero exit code.
+local binary=""
+local broken_binaries=""
+# Catch all binaries and libraries also e.g. those that are copied via COPY_AS_IS into other paths:
+for binary in $( find $ROOTFS_DIR -type f -executable -printf '/%P\n' ) ; do
+    chroot $ROOTFS_DIR /bin/ldd $binary | grep -q 'not found' && broken_binaries="$broken_binaries $binary"
```

Comment:
For binaries copied by `COPY_AS_IS` into paths with blanks this will not fail but produce misleading output.

If we could rely on Bash 4 and `readarray` those sort of things would be really really easy. With Bash 3 using arrays means more manual coding work. It would also be more like the other ReaR code.

Crazy idea: Maybe we just need to define a `readarray` function if ReaR runs on older systems and then we can use it.

Reply:
I played a bit with a binary with a blank "/bin/my prog"
and found no sufficiently easy way how to deal with it:
<pre>
# ls /tmp/rear.yH2NEsUbyyDbKRT/rootfs/bin/my*
/tmp/rear.yH2NEsUbyyDbKRT/rootfs/bin/my prog

# for binary in $( find /tmp/rear.yH2NEsUbyyDbKRT/rootfs/bin -type f -executable -printf '/%P\n' ) ; do echo "'$binary'" ; done | grep -2 my
'/sort'
'/rpcinfo'
'/my'
'prog'
'/sg_inq'

# bins=( $( find /tmp/rear.yH2NEsUbyyDbKRT/rootfs/bin -type f -executable -printf '/%P\n' ) )

# for b in "${bins[@]}" ; do echo "'$b'" ; done | grep -2 my
'/sort'
'/rpcinfo'
'/my'
'prog'
'/sg_inq'

# ( IFS=$'\n' ; for binary in $( find /tmp/rear.yH2NEsUbyyDbKRT/rootfs/bin -type f -executable -printf '/%P\n' ) ; do echo "'$binary'" ; done | grep -2 my )
'/sort'
'/rpcinfo'
'/my prog'
'/sg_inq'
'/netstat'
</pre>
But somehow I would really prefer not to have
to mess around with IFS...

Reply:
Because file names can containin newline characters
any messing around with IFS will fail for exactly
those character(s) that are used as IFS.